README.md
# Sundew
WMO-GTS switch on TCP/IP protocols (largely obsolete, legacy code.)

History:

This was started in 2004, entered full production 2007.  about a dozen different developers over it's life.
This application is 25 klocs of python, that replaced Tandem/Apps 450 klocs + PDS (45 klocs) (20:1) reduction
in application code.  Provided most Meteorological switching for Canada from about 2006 until about 2017.  
Still in service but being gradually supplanted by Sarracenia.  Same code base support both message and
file transfer (was innovative at the time.)

Sundew is too constrained by early requirements to be modernized, so started a fresh code base
to work with modern needs: Sarracenia will eventually replace Sundew for (modernized versions of)
the same use case.

Suggest you look there if looking for something new to deploy: http://MetPX.github.com/sarracenia

Features:

Implements the WMO-GTS TCP/IP socket protocol (which is now deprecated by WMO.)
Does not implement any X.25 (also deprecated by WMO.)
Does implement a specific implementation of AFTN over TCP/IP.
Does not handle dialup data collection (seems many expect a *wmo switch* to do that.)
Whatever products it acquires, it stores it in a file, and then links the file to queue directories to be sent
on to a destination based on a routing table. 




RELEASE_NOTES

---------------- Below are changes made up to version px-0-6-1 --------------------------------------

2013-09      (MG) clean up trunk to restart at px-0-6-0

             (MG) Client.py
                  execfile name rearranged
                  AMQP options revisited

                  Logger.py
                  addition of class LoggerMonitor
                  insert module start_timer and delivered 
                  for logging of latency and speed for senders

                  SenderFTP.py
                  sftp  rename for .tmp lock uncommented
                  sftp  supports throttling
                  execfile name rearranged
                  use logger.start_timer() and logger.delivered for latency and speed 

                  Source.py
                  AMQP options revisited

                  receiverAmqp.py
                  AMQP options revisited
                  arguments in function calls revisited
                  Use ingestor instead of bulletin manager 
                  Use message properties application_headers for filename
                  Exchange auto-delete set to False

                  senderAMIS.py
                  use logger.start_timer() and logger.delivered for latency and speed 

                  senderAMQP.py
                  AMQP options revisited
                  arguments in function calls revisited
                  Use message properties application_headers for filename
                  use logger.start_timer() and logger.delivered for latency and speed 
                  Exchange auto-delete set to False

                  senderAm.py
                  use logger.start_timer() and logger.delivered for latency and speed 
                  execfile name rearranged

                  senderWmo.py
                  use logger.start_timer() and logger.delivered for latency and speed 

                  DirectRoutingParser.py
                  option include "filename" now supported


---------------- Below are changes made up to version px-0-6-0 --------------------------------------

*** Note  CMO packaging 

2012-03-02   (JH) lib/LatMessage.py lib/Latencies.py lib/PDSLatencies.py lib/PXLatencies.py
                  lib/Plotter.py lib/copyFiles.py bin/paplat   change python2 for python

             (JH) rename .localRetrans for localRetrans and modifi pxRetrans accordingly
                  rename hang-check.sh for hang-check

             (JH) move doc from trunk to sundew
                  debian/metpx-sundew.docs

             (JH) debian/changelog modified
                  debian/control   modified for architecture and dependencies
                  debian/copyright updated
                  debian/postinst  PXHOME from getent and use dpkg-statoverride for directories
                  debian/rules     modified to install doc (exclude .svn)
                                   create queuedirs and db dir
                                   move  dh_compress for the man

             (JH) /etc/init.d/px written and packaged

             (MG) bulletinLib.py utils.py deleted
                  short manpages written for  pxValidator, pxRetrans, pxFlowNames, pxDBCleaner,
                  pxChecker, mr-clean, localRetrans, hang-check, check-unique, paplat


---------------- Below are changes made up to version px-0-5-9 --------------------------------------

2012-01-17   (MG) Ingestor.py Attempt to correct name clash in db by adding a second
                  to the date suffix 

2011-12-15   (MG) DirectRoutingParser.py bug fix : return priority  when product match key_accept 

2011-11-18   (MG) SenderScript.py  client object first argument of send_script

2011-10-26   (MG) Ingestor.py Routingtable can be modified for bulletins too.

2011-10-04   (MG) Source.py/CollectionBuilder.py moved aaxx, metar and taf as receiver options

2011-10-04   (MG) StationParser.py trap error and log when a station entry is wrong and continue

2011-08-27   (DL) pxRetrans Option waitTimeForLogEntry : confirms resending within n sec from log

2011-07-27   (MG) PXIgnitor.py   kill sender if stop did not work after .5 sec.

2011-01-20   (MG) Source.py/Ingestor.py add pull-script capability

2010-11-19   (MG) PullFTP.py  added pattern match like JJJ julian day and JJJ-1...



lib/Client.py
"""
MetPX Copyright (C) 2004-2006  Environment Canada
MetPX comes with ABSOLUTELY NO WARRANTY; For details type see the file
named COPYING in the root of the source directory tree.
"""

"""
#############################################################################################
# Name: Client.py
#
# Authors: Peter Silva (imperative style)
#          Daniel Lemay (OO style)
#          Michel Grenier (directory path with patterns)
#                         (directory path mkdir if doesn't exist)
#                         (maxLength for segmentation when needed)
#
# Date: 2005-01-10 (Initial version by PS)
#       2005-08-21 (OO version by DL)
#       2005-11-01 (Path stuff by MG)
#
# Description:
#
# MG python3 compatible
#
#############################################################################################

"""
import sys, os, re, time, fnmatch
import PXPaths
from URLParser import URLParser
from Logger import Logger
#from Flow import Flow

PXPaths.normalPaths()              # Access to PX paths

class Client(object):

    def __init__(self, name='toto', logger=None) :

        #Flow.__init__(self, name, 'sender', type, batch) # Parent constructor

        # General Attributes
        self.name = name                          # Client's name
        if logger is None:
            self.logger = Logger(PXPaths.LOG + 'tx_' + name + '.log', 'INFO', 'TX' + name) # Enable logging
            self.logger = self.logger.getLogger()
        else:
            self.logger = logger
        self.logger.info("Initialisation of client %s" % self.name)
        self.debug = False                        # If we want sections with debug code to be executed
        self.host = 'localhost'                   # Remote host address (or ip) where to send files
        self.type = 'single-file'                 # Must be in ['single-file', 'bulletin-file', 'file', 'am', 'wmo', 'amis']
        self.protocol = None                      # First thing in the url: ftp, file, am, wmo, amis
        self.batch = 100                          # Number of files that will be read in each pass
        self.kbytes_ps = -1.0                     # Number of Kbytes per seconds targetted
        self.cache_size = 125000                  # Maximum Number of md5sum from files kept in cache manager
        self.timeout = 10                         # Time we wait between each tentative to connect
        self.maxLength = 0                        # max Length of a message... limit use for segmentation, 0 means unused

        self.validation = True                    # Validation of the filename (prio + date)

        self.purge = None                         # Purge instructions (ex: 7H,4+:10H,3)
        self.purgeAliases = {}                    # Purge Aliases
        #self.purgeAliases['TEST'] = '8H,4+:18H,3' # Purge priority 4,5 older than 10 hours and priority 3 older than 18 hours

        self.purgeInst = []                       # Purge instructions parsed 

        self.patternMatching = True               # Verification of the accept and reject mask of the client before sending a file
        self.nodups = True                        # Check if the file has already been sent (md5sum present in the cache)
        self.mtime = 0                            # Integer indicating the number of seconds a file must not have
                                                  # been touched before being picked

        self.sorter = 'MultiKeysStringSorter'     # Class (or object) used to sort
        self.masks = []                           # All the masks (accept and reject)
        self.masks_deprecated = []                # All the masks (imask and emask)
        self.url = None
        self.collection = None                    # Client do not participate in the collection effort

        # Socket Attributes
        self.port = None                    # Port number
        self.keepAlive = True               # TCP SO_KEEPALIVE on (True) or off(False)

        # Am destination thread Attributes
        self.am_dest_thread = None                # assign a destination thread number to a specific bulletin header

        # Files Attributes
        self.user = None                    # User name used to connect
        self.passwd = None                  # Password 
        self.ssh_keyfile = None             # ssh private key file for sftp protocol
        self.ftp_mode = 'passive'           # Default is 'passive', can be set to 'active'
        self.dir_mkdir = False              # Verification and creation of directory inside ftp...
        self.dir_pattern = False            # Verification of patterns in destination directory
        self.binary = True                  # FTP uses binary mode 

        self.chmod = 666                    # when the file is delevered chmod it to this value
        self.timeout_send = 0               # Timeout in sec. to consider a send to hang ( 0 means inactive )
        self.lock = '.tmp'                  # file send with extension .tmp for lock
                                            # if lock == "umask" than use umask 777 to put files


        self.destfn_script   = None         # a script to rename the file for client
        self.destfn_execfile = None

        self.dx_script       = None         # a script to convert the data of a senderAm
        self.dx_execfile     = None

        self.fx_script       = None         # a script to convert the file for client
        self.fx_execfile     = None

        self.send_script     = None         # a script to send a list of files
        self.send_execfile   = None

        # AMQP

        self.exchange_content = 'text/plain'
        self.exchange_key     = ''
        self.exchange_name    = None
        self.exchange_realm   = '/data'
        self.exchange_type    = 'fanout'

        # All defaults for a source were set earlier in this class
        # But some of them may have been overwritten in the px.conf file
        # Load the px.conf stuff related to the source

        pxconf_Path = PXPaths.ETC + 'px.conf'
        if os.path.isfile(pxconf_Path) : self.readConfigFile( pxconf_Path, '.', 'WHATFN' )

        # read in the config

        self.readConfig()

        if self.destfn_execfile != None :
           try    : self.execfile(PXPaths.SCRIPTS + self.destfn_execfile )
           except : self.logger.error("Problem with destfn_script %s" % self.destfn_execfile)

        if self.fx_execfile != None :
           try    : self.execfile(PXPaths.SCRIPTS + self.fx_execfile )
           except : self.logger.error("Problem with fx_script %s" % self.fx_execfile)

        if self.dx_execfile != None :
           try    : self.execfile(PXPaths.SCRIPTS + self.dx_execfile )
           except : self.logger.error("Problem with dx_script %s" % self.dx_execfile)

        if self.send_execfile != None :
           try    : self.execfile(PXPaths.SCRIPTS + self.send_execfile )
           except : self.logger.error("Problem with send_script %s" % self.send_execfile)

        #self.printInfos(self)


    def execfile(self, path):
        exec(compile(open(path).read(), path, 'exec'))

    def parsePurgeInstructions(self, line):
        # '10H,4+:18H,3'
        MINUTE = 60
        HOUR = 60 * MINUTE

        newInst = []

        instructions = line.strip().split(':')
        for instruction in instructions:
            hours, pri = instruction.strip().split(',')    
            hours.strip(), pri.strip()
            sec = int(hours[:-1]) * HOUR
            if pri[-1] == '+':
                pri = list(range(int(pri[:-1]), 6))
            elif int(pri) in [1,2,3,4,5]:
                pri = [int(pri)]
            newInst.append((sec,pri))

        self.purgeInst = newInst

    def readConfig(self):
        currentDir = '.'                # Current directory
        currentFileOption = 'WHATFN'    # Under what filename the file will be sent (WHATFN, NONE, etc., See PDS)
        fileName = PXPaths.TX_CONF +  self.name + '.conf'
        #print filePath
        self.readConfigFile(fileName,currentDir,currentFileOption)

    def readConfigFile(self,filePath,currentDir,currentFileOption):
        
        def isTrue(s):
            if  s == 'True' or s == 'true' or s == 'yes' or s == 'on' or \
                s == 'Yes' or s == 'YES' or s == 'TRUE' or s == 'ON' or \
                s == '1' or  s == 'On' :
                return True
            else:
                return False

        def stringToOctal(string):
            if len(string) != 3:
                return 0o644
            else:
                return int(string[0])*64 + int(string[1])*8 + int(string[2])

        try:
            config = open(filePath, 'r')
        except:
            (type, value, tb) = sys.exc_info()
            print("Type: %s, Value: %s" % (type, value))
            return 

        for line in config.readlines():
            words = line.split()
            if (len(words) >= 2 and not re.compile('^[ \t]*#').search(line)):
                try:
                    if   words[0] == 'accept':
                         cmask       = re.compile(words[1])
                         cFileOption = currentFileOption
                         if len(words) > 2: cFileOption = words[2]
                         self.masks.append((words[1], currentDir, cFileOption, cmask, True))
                    elif words[0] == 'reject':
                         cmask = re.compile(words[1])
                         self.masks.append((words[1], currentDir, currentFileOption, cmask, False))

                    elif words[0] == 'imask': self.masks_deprecated.append((words[1], currentDir, currentFileOption))  
                    elif words[0] == 'emask': self.masks_deprecated.append((words[1],))

                    elif words[0] == 'directory': currentDir = words[1]
                    elif words[0] == 'filename': currentFileOption = words[1]
                    elif words[0] == 'include':
                        fileName = PXPaths.TX_CONF + words[1]
                        self.readConfigFile(fileName,currentDir,currentFileOption)
                    elif words[0] == 'destination':
                        self.url = words[1]
                        urlParser = URLParser(words[1])
                        (self.protocol, currentDir, self.user, self.passwd, self.host, self.port) =  urlParser.parse()
                        if len(words) > 2:
                            currentFileOption = words[2]
                    elif words[0] == 'validation': self.validation =  isTrue(words[1])
                    elif words[0] == 'purgeAlias':
                        self.purgeAliases[words[1]] = words[2]
                    elif words[0] == 'purge':
                        try:
                            self.purge = self.purgeAliases[words[1]]
                            self.parsePurgeInstructions(self.purgeAliases[words[1]])
                        except:
                            self.purge = words[1]
                            self.parsePurgeInstructions(words[1])

                    elif words[0] == 'noduplicates': self.nodups =  isTrue(words[1])
                    elif words[0] == 'patternMatching': self.patternMatching =  isTrue(words[1])
                    elif words[0] == 'keepAlive': self.keepAlive =  isTrue(words[1])
                    elif words[0] == 'mtime': self.mtime = int(words[1])
                    elif words[0] == 'sorter': self.sorter = words[1]
                    elif words[0] == 'type': self.type = words[1]
                    elif words[0] == 'protocol': self.protocol = words[1]
                    elif words[0] == 'maxLength': self.maxLength = int(words[1])
                    elif words[0] == 'host': self.host = words[1]
                    elif words[0] == 'port': self.port = int(words[1])
                    elif words[0] == 'user': self.user = words[1]
                    elif words[0] == 'password': self.passwd = words[1]
                    elif words[0] == 'ssh_keyfile': self.ssh_keyfile = words[1]
                    elif words[0] == 'batch': self.batch = int(words[1])
                    elif words[0] == 'kbytes_ps': self.kbytes_ps = int(words[1]) + 0.0
                    elif words[0] == 'cache_size': self.cache_size = int(words[1])
                    elif words[0] == 'debug' and isTrue(words[1]): self.debug = True
                    elif words[0] == 'timeout': self.timeout = int(words[1])
                    elif words[0] == 'chmod': self.chmod = int(words[1])
                    elif words[0] == 'timeout_send': self.timeout_send = int(words[1])
                    elif words[0] == 'lock': self.lock = words[1]
                    elif words[0] == 'ftp_mode': self.ftp_mode = words[1]
                    elif words[0] == 'dir_pattern': self.dir_pattern =  isTrue(words[1])
                    elif words[0] == 'binary': self.binary =  isTrue(words[1])
                    elif words[0] == 'dir_mkdir': self.dir_mkdir =  isTrue(words[1])
                    elif words[0] == 'destfn_script': self.destfn_execfile = words[1]
                    elif words[0] == 'dx_script': self.dx_execfile = words[1]
                    elif words[0] == 'fx_script': self.fx_execfile = words[1]
                    elif words[0] == 'send_script': self.send_execfile = words[1]

                    elif words[0] == 'am_dest_thread':
                         if self.am_dest_thread == None : self.am_dest_thread = {}
                         self.am_dest_thread[words[1]] = int(words[2])

                    # AMQP
                    elif words[0] == 'exchange_content': self.exchange_content = words[1]
                    elif words[0] == 'exchange_key': self.exchange_key = words[1]
                    elif words[0] == 'exchange_name': self.exchange_name = words[1]
                    elif words[0] == 'exchange_realm': self.exchange_realm = words[1]
                    elif words[0] == 'exchange_type': 
                         if words[1] in ['fanout','direct','topic','headers'] :
                            self.exchange_type = words[1]
                         else :
                            self.logger.error("Problem with exchange_type %s" % words[1])

                except:
                    self.logger.error("Problem with this line (%s) in configuration file of client %s" % (words, self.name))

        #if not self.validation:
        #    self.sorter = 'None'    # Must be a string because eval will be subsequently applied to this

        config.close()
    
        #self.logger.debug("Configuration file of client %s has been read" % (self.name))

    def run_destfn_script(self, filename): 
        if self.destfn_script == None : return filename
        return self.destfn_script(filename)

    def run_dx_script(self, data, logger): 
        if self.dx_script == None : return data
        return self.dx_script(data, logger)

    def run_fx_script(self, filename, logger): 
        if self.fx_script == None : return filename
        return self.fx_script(filename, logger)

    def _getMatchingMask(self, filename): 

        # check for deprecated use
        if len(self.masks_deprecated) > 0 :
           for mask in self.masks_deprecated:
               if fnmatch.fnmatch(filename, mask[0]):
                  try:
                          if mask[2]:
                             return mask
                  except:
                          return None

        # new regexp mask usage
        for mask in self.masks:
            if mask[3].match(filename) :
               if mask[4] : return mask
               return None

        return None

    def getDestInfos(self, filename):
        """
        WHATFN         -- First part (':') of filename 
        HEADFN         -- Use first 2 fields of filename
        NONE           -- Use the entire filename
        TIME or TIME:  -- TIME stamp appended
        DESTFN=fname   -- Change the filename to fname

        ex: mask[2] = 'NONE:TIME'
        """
        mask = self._getMatchingMask(filename)
        if mask:
            timeSuffix   = ''
            satnet       = ''
            parts        = filename.split(':')
            firstPart    = parts[0]
            destFileName = filename
            for spec in mask[2].split(':'):
                if spec == 'WHATFN':
                    destFileName =  firstPart
                elif spec == 'HEADFN':
                    headParts = firstPart.split('_')
                    if len(headParts) >= 2:
                        destFileName = headParts[0] + '_' + headParts[1] 
                    else:
                        destFileName = headParts[0] 
                elif spec == 'SENDER':
                    if len(parts[4]) == 1 : sender = parts[6]
                    else                  : sender = parts[5]
                    if sender[:6] == 'SENDER' : 
                       destFileName = sender.split('=')[1] 
                elif spec == 'NONE':
                    # PX default behavior : keep 6 first fields
                    destFileName = ':'.join(parts[:6])
                    #  PDS default behavior  keep 5 first fields
                    if len(parts[4]) != 1 : destFileName = ':'.join(parts[:5])
                    # extra trailing : removed if present
                    if destFileName[-1] == ':' : destFileName = destFileName[:-1]
                elif spec == 'NONESENDER':
                    # same as NONE
                    destFileName = ':'.join(parts[:6])
                    if len(parts[4]) != 1 : destFileName = ':'.join(parts[:5])
                    if destFileName[-1] == ':' : destFileName = destFileName[:-1]
                    # add SENDER if present
                    sender = parts[6]
                    if len(parts[4]) != 1 : sender = parts[5]
                    if len(sender) > 6 and sender[:6] == 'SENDER' : 
                       destFileName = destFileName + ':' + sender
                elif re.compile('SATNET=.*').match(spec):
                    satnet = ':' + spec
                elif re.compile('DESTFN=.*').match(spec):
                    destFileName = spec[7:]
                elif re.compile('DESTFNSCRIPT=.*').match(spec):
                     old_destfn_script = self.destfn_script
                     old_destFileName  = destFileName
                     script = PXPaths.SCRIPTS + spec[13:]
                     try    :
                              self.execfile(script)
                              destFileName = self.destfn_script(filename)
                     except : self.logger.error("Problem with filename DESTFNSCRIPT=%s" % script)
                     self.destfn_script = old_destfn_script
                     if destFileName == None : destFileName = old_destFileName
                elif spec == 'TIME':
                    if destFileName != filename :
                       timeSuffix = ':' + time.strftime("%Y%m%d%H%M%S", time.gmtime())
                       # check for PX or PDS behavior ... if file already had a time extension keep his...
                       if parts[-1][0] == '2' : timeSuffix = ':' + parts[-1]
                else:
                    self.logger.error("Don't understand this DESTFN parameter: %s" % spec)
                    return (None, None) 
            return (destFileName + satnet + timeSuffix, mask[1])
        else:
            return (None, None) 

    def printInfos(self, client):
        print("==========================================================================")
        print("Name: %s " % client.name)
        print("Host: %s" % client.host)
        print("Type: %s" % client.type)
        print("Protocol: %s" % client.protocol)
        print("Batch: %s" %  client.batch)
        print("Kbytes_ps: %s" %  client.kbytes_ps)
        print("Cache size: %s" %  client.cache_size)
        print("Max length: %i" % client.maxLength)
        print("Mtime: %i" % client.mtime)
        print("Timeout: %s" % client.timeout)
        print("Sorter: %s" % client.sorter)
        print("URL: %s" % client.url)
        print("Port: %s" % client.port)
        print("User: %s" % client.user)
        print("Passwd: %s" % client.passwd)
        print("ssh_keyfile: %s" % client.ssh_keyfile)
        print("Chmod: %s" % client.chmod)
        print("TCP SO_KEEPALIVE: %s" % client.keepAlive)
        print("Timeout_send: %i" % client.timeout_send)
        print("Lock: %s" % client.lock)
        print("FTP Mode: %s" % client.ftp_mode)
        print("DIR Pattern: %s" % client.dir_pattern)
        print("binary : %s" % client.binary)
        print("DIR Mkdir  : %s" % client.dir_mkdir)
        print("Validation: %s" % client.validation)
        print("Purge Aliases: %s" % client.purgeAliases)
        print("Purge: %s" % client.purge)
        print("purgeInst: %s" % client.purgeInst)
        print("Pattern Matching: %s" % client.patternMatching)
        print("No duplicates: %s" % client.nodups)
        print("am_dest_thread: %s" % client.am_dest_thread)

        print("******************************************")
        print("*       AMQP stuff                       *")
        print("******************************************")

        print("exchange_content: %s" % client.exchange_content)
        print("exchange_key: %s" % client.exchange_key)
        print("exchange_name: %s" % client.exchange_name)
        print("exchange_type: %s" % client.exchange_type)
        print("exchange_realm: %s" % client.exchange_realm)

        print("******************************************")
        print("*       Client Masks                     *")
        print("******************************************")

        for mask in self.masks:
            if mask[4] :
               print(" accept %s" % mask[0] )
            else :
               print(" reject %s" % mask[0] )

        print("******************************************")
        print("*       Client Masks                     *")
        print("******************************************")

        for mask in self.masks_deprecated:
            print(" deprecated %s =" % mask[0] )
        print("==========================================================================")

if __name__ == '__main__':

    """
    client =  Client('wxo-b1')
    client.readConfig()
    client.printInfos(client)
    print client.getDestInfos('AWCNALO:TUTU:TITI:TOTO:MIMI:Directi')
    print client.getDestInfos('FPCN_DAN_MAN_lkdslfk:TUTU:TITI:TOTO:MIMI:Directi')
    print client.getDestInfos('WLCN_MAN_lkdslfk:TUTU:TITI:TOTO:MIMI:Direct')
    print client.getDestInfos('SMALLO:TUTU:TITI:TOTO:MIMI:Direct')
    print client.getDestInfos('WTCNALLO:TUTU:TITI:TOTO:MIMI:Direct')

    """
    for filename in os.listdir(PXPaths.TX_CONF):
        print(filename)
        if filename[-5:] != '.conf': 
            continue
        else:
            client = Client(filename[0:-5])
            client.readConfig()
            client.printInfos(client)

